### PR TITLE
Fix lakectl local commits remote changes outside synced prefix

### LINKED ISSUE #7687: [Bug]: `lakectl local commit` commits staged changes outside of its directory
### What happened?

What actually happened, including error codes if applies.
        
Steps to Reproduce:
1. Ensure lakefs://repo/main is _dirty_ but has no "directory" path `xyzzy`.
1. `lakectl local init lakefs://repo/main/xyzzy`
1. `touch x`
1. `lakectl local commit --message 'add xyzzy for lulz'`
1. `lakectl diff lakefs://repo/main{~,}`:

   ```
   Left ref: lakefs://ariels-repo/main~
   Right ref: lakefs://ariels-repo/main
   + added .hidden-foo
   + added foofoo
   + added xyzzy/x
   ```

These 2 files were added by `lakectl local` because _I added them elsewhere_.

### Expected behavior

`lakectl commit` should fail with an error about the branch being dirty, or lakectl should be more like Git LFS and not hide changes from me.

### lakeFS version

_No response_

### How lakeFS is installed

_No response_

### Affected clients

_No response_

### Relevant log output

_No response_

### Contact details

_No response_

### PR DESCRIPTION
**Pull request by eladlachmi**
Closes #7687 

#### The Problem

Please see the associated issue for the full description of the problem.

#### Root Cause

lakeFS, unlike Git, doesn't support the concept of committing only staged changes. A commit operation will commit any uncommitted diffs between the refs or fail if conflicts occur (all or nothing).  
`lakectl local` isn't special in this regard. However, being able to sync with any arbitrary prefix creates the expectation that any operations will be scoped to that prefix. In practice, this isn't the case. Merging a ref in `lakectl local` unexpectedly merges all changes in the remote ref, regardless of prefix.

#### Solution

This PR adds to `lakectl local commit` a check for existing uncommitted changes on the remote ref, which are outside of the synced prefix. If such changes are found, the operation is aborted and an informative error message is shown. This behavior can be overridden by adding the the `--force` flag to the commit command.
      
### Testing Details

These changes were tested manually and with unit tests included in this PR.  

**Note to reviewers:** I believe the included tests cover the possible cases. I would like your feedback if you think that any additional tests are needed for better coverage.

### Breaking Change?

**This is a breaking change!**  
Although this behavior is unexpected and considered a bug, it will change following this PR. `lakectl local` users relying on this behavior will see an error message instead of the operation completed successfully and will need to add the `--force` flag to get back the previous behavior.

Special thanks to @arielshaqed for the refresher course on `char` ordinals and string sort order 😅👑

### CODE REVIEW

**Comment by N-o-Z** [comment]
@eladlachmi I honestly think this issue was mislabled as bug. This is a feature request and regardless it constitutes as a breaking change

File: cmd/lakectl/cmd/local_commit.go [hunk 1]
```diff
@@ -24,6 +32,47 @@ func findConflicts(changes local.Changes) (conflicts []string) {
...
+		return false
+	}
+	if slices.ContainsFunc(dirtyResp.JSON200.Results, func(diff apigen.Diff) bool {
+		return diff.PathType == "object" && !strings.HasPrefix(diff.Path, *currentURI.Path)
+	}) {
+		return true
+	}
+
+	// Get the first uncommitted change after the prefix. If it exists, we're also done
+	nextPrefix := fmt.Sprintf("%s%s", filepath.Clean(*currentURI.Path), asciiCharAfterSlash)
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/local_commit.go:57]: I'm not clear what we are doing here. We should optimally get the next prefix in line and perform the diff on that.
I don't understand how the statObject will help, we're assuming some object with some path will exist?
Overall this is very confusing and feels over-complicated

File: cmd/lakectl/cmd/local_commit.go [hunk 2]
```diff
@@ -24,6 +32,47 @@ func findConflicts(changes local.Changes) (conflicts []string) {
...
+	dirtyResp, err = client.DiffBranchWithResponse(ctx, remote.Repository, remote.Ref, &apigen.DiffBranchParams{
+		Amount: apiutil.Ptr(apigen.PaginationAmount(1)),
+		After:  apiutil.Ptr(apigen.PaginationAfter(nextPrefix)),
+	})
+	DieOnErrorOrUnexpectedStatusCode(dirtyResp, err, http.StatusOK)
+
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/local_commit.go:63]: We can stop here instead of creating an unnecessary API call

File: cmd/lakectl/cmd/local_commit.go [hunk 3]
```diff
@@ -24,6 +32,47 @@ func findConflicts(changes local.Changes) (conflicts []string) {
...
+	// The above gives us SeekGT. Since we need SeekGE, we do another stat for exact match
+	statResp, err := client.StatObjectWithResponse(ctx, remote.Repository, remote.Ref, &apigen.StatObjectParams{
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/local_commit.go:65]: Feels like a shot in the dark

File: cmd/lakectl/cmd/local_commit.go [hunk 4]
```diff
@@ -56,6 +106,11 @@ var localCommitCmd = &cobra.Command{
 		resp, err := client.GetBranchWithResponse(cmd.Context(), remote.Repository, remote.Ref)
 		DieOnErrorOrUnexpectedStatusCode(resp, err, http.StatusOK)
 
+		hasChangesOutsideSyncedPrefix := hasUncommittedOutsideLocalPrefix(cmd.Context(), client, remote, idx)
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/local_commit.go:109]:
```suggestion
		hasExternalChange := hasUncommittedOutsideLocalPrefix(cmd.Context(), client, remote, idx)
```

File: cmd/lakectl/cmd/local_commit.go [hunk 5]
```diff
@@ -56,6 +106,11 @@ var localCommitCmd = &cobra.Command{
...
+		if hasChangesOutsideSyncedPrefix && !force {
+			DieFmt("The branch you are trying to commit to contains uncommitted changes outside the lakeFS path your local directory '%s' is synced with.\nTo proceed, use the --force flag.", localPath)
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/local_commit.go:111]: You already know the name of the branch:

```suggestion
			DieFmt("Branch %s contains uncommitted changes outside of local path '%s'.\nTo proceed, use the --force flag.", remote.ref, localPath)
```

File: cmd/lakectl/cmd/local_commit.go [hunk 6]
```diff
@@ -170,6 +225,7 @@ var localCommitCmd = &cobra.Command{
 
 //nolint:gochecknoinits
 func init() {
+	withForceFlag(localCommitCmd, "Commit changes even if remote branch includes uncommitted changes outside the synced directory")
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/local_commit.go:228]:
```suggestion
	withForceFlag(localCommitCmd, "Commit changes even if remote branch includes uncommitted changes external to the synced path")
```

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Thanks, really like the approach.
Regarding the tests - I like the unit tests but I would honestly just replace them with a scenario in esti - as we already have tests for lakectl local + commit. I would just add the relevant scenarios there

**Comment by eladlachmi** [comment]
> @eladlachmi I honestly think this issue was mislabled as bug. This is a feature request and regardless it constitutes as a breaking change

@N-o-Z IMO, any surprising/unexpected behavior hidden from or not clearly communicated to the user is a bug.  

I also think we're using the term "breaking change" too loosely. This isn't the same thing as a breaking change in an API. 
This is more like a change in the UI.  

For example, changing the default checked/unchecked state of a checkbox in a modal is undoubtedly a missable change, and you can and should draw attention clearly to it, at least for the next few versions. However, unlike an API, this is user-interactive, and the option to revert to the previous behavior is just a `--force` flag away.  

I agree that we need to clearly mark this in the release notes as a change `lakectl local` users need to be aware of, but the term "breaking change" has ramifications, which I don't think apply in this case.

Differing opinions aside, beyond clearly marking and communicating that this is a behavior change, is there anything else you think we need to do before, during, or after releasing this change?

(p.s. to spare the need to add `--force` each time for users who want this behavior, maybe we can add this as a config option as well? WDYT?)

**Comment by eladlachmi** [comment]
> Thanks, really like the approach. Regarding the tests - I like the unit tests but I would honestly just replace them with a scenario in esti - as we already have tests for lakectl local + commit. I would just add the relevant scenarios there

@N-o-Z I don't mind adding esti tests, but since the unit tests are already written, and they provide a localized and fast feedback loop (unlike esti tests), I don't see a reason to remove them. I think a multi-layer testing approach provides good coverage + a good developer experience. WDYT?

File: cmd/lakectl/cmd/local_commit.go [hunk 7]
```diff
@@ -24,6 +32,47 @@ func findConflicts(changes local.Changes) (conflicts []string) {
 	return
 }
 
+func hasUncommittedOutsideLocalPrefix(ctx context.Context, client *apigen.ClientWithResponses, remote *uri.URI, idx *local.Index) bool {
```
[the remark this answers is not in the document]
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go:35]: I share the credit with @arielshaqed 😁

**Comment by N-o-Z** [comment]
> > @eladlachmi I honestly think this issue was mislabled as bug. This is a feature request and regardless it constitutes as a breaking change
> 
> @N-o-Z IMO, any surprising/unexpected behavior hidden from or not clearly communicated to the user is a bug.
> 
> I also think we're using the term "breaking change" too loosely. This isn't the same thing as a breaking change in an API. This is more like a change in the UI.
> 
> For example, changing the default checked/unchecked state of a checkbox in a modal is undoubtedly a missable change, and you can and should draw attention clearly to it, at least for the next few versions. However, unlike an API, this is user-interactive, and the option to revert to the previous behavior is just a `--force` flag away.
> 
> I agree that we need to clearly mark this in the release notes as a change `lakectl local` users need to be aware of, but the term "breaking change" has ramifications, which I don't think apply in this case.
> 
> Differing opinions aside, beyond clearly marking and communicating that this is a behavior change, is there anything else you think we need to do before, during, or after releasing this change?
> 
> (p.s. to spare the need to add `--force` each time for users who want this behavior, maybe we can add this as a config option as well? WDYT?)

Breaking change doesn't necessarily means we need to bump a major version and we shouldn't be afraid to use the term. Simply communicate it appropriately and verify this is understood by all stakeholders (@treeverse/product FYI).
I reject the comparison with UI changes - this is clearly a change in logic. Theoretically users who have used this feature as is and might have also relied on the current behavior might find their flows and automation broken. I'm not familiar of any user who automates flow via UI 😵‍💫 
Definitely go with the `force` flag - I see no justification to add it to the configuration

File: cmd/lakectl/cmd/local_commit.go (hunk 1 above)
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go:57]: Ok, so I'll explain the approach. Let's say I synced locally with the prefix `test-data/`.

Case 1: There are no uncommitted changes. The first check doesn't return any results. We're done—continue with the commit.

Case 2: We have uncommitted changes _before_ `test-data/`, e.g., `abc/file.parquet`. The first check will return a result outside the prefix. We're done—exit with an error message.

Case 3: We have uncommitted changes _after_ `test-data/`, e.g., `zzz/file.parquet`. The second check needs to fetch the first uncommitted change after `test-data/`. In lexicographic order, this is `test-data0` because `/` is ordinal 47 and the next ordinal (48) is `0`. This fetch will return a result-exit with an error message.

These three cases seem to cover all the options. However, there's a subtle edge case. Our `After` parameter acts like `SeekGT`. This means that we've covered everything up to `test-data0` (end not inclusive) and everything from `test-data0` (start not inclusive). So if we happen to have an object specifically named `test-data0`, we'll miss it. That's what the last `StatObject` call is for.

Hope this helps clear things up 😊

File: cmd/lakectl/cmd/local_commit.go (hunk 2 above)
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go:63]: See the explanation in the previous comment's reply and let me know what you think.

File: cmd/lakectl/cmd/local_commit.go (hunk 3 above)
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go:65]: See the explanation in the previous comment's reply and let me know what you think 😅

[the remark this answers is not in the document]
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go:65]: Very true. Fixed.

File: cmd/lakectl/cmd/local_commit.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go]: Done.

File: cmd/lakectl/cmd/local_commit.go (hunk 4 above)
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go]: Done.

File: cmd/lakectl/cmd/local_commit.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go]: As you know, concise is my middle name 😜

File: cmd/lakectl/cmd/local_commit.go (hunk 5 above)
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go]: Done.

File: cmd/lakectl/cmd/local_commit.go (hunk 6 above)
Comment (eladlachmi) [reply cmd/lakectl/cmd/local_commit.go]: Done.

**Comment by N-o-Z** [comment]
> > Thanks, really like the approach. Regarding the tests - I like the unit tests but I would honestly just replace them with a scenario in esti - as we already have tests for lakectl local + commit. I would just add the relevant scenarios there
> 
> @N-o-Z I don't mind adding esti tests, but since the unit tests are already written, and they provide a localized and fast feedback loop (unlike esti tests), I don't see a reason to remove them. I think a multi-layer testing approach provides good coverage + a good developer experience. WDYT?

IMO, mocking the server does not help us identify changes in behavior vs the actual server (extra important since lakectl and lakeFS are different binaries) so we have to have esti test. Once we have this covered in esti anyway, the unit tests become pretty redundant. I don't mind leaving them but I have to insist on having the integration tests

**Comment by eladlachmi** [comment]
>I'm not familiar of any user who automates flow via UI 😵‍💫

@N-o-Z How confident are you about your coverage? 😅  

Anyway... I think both the CLI and UI are user-interactive tools that users can choose to automate, but in doing so, they should be aware that interfaces (command line or otherwise) don't carry the same guarantees as APIs and SDKs.  
Let's agree to disagree 🤝  

As long as we agree about what we intend to do, we're good.

File: esti/lakectl_local_test.go [hunk 8]
```diff
@@ -464,6 +464,145 @@ func TestLakectlLocal_commit(t *testing.T) {
...
+			if len(tc.uncommittedRemote) > 0 {
+				for _, f := range tc.uncommittedRemote {
+					vars["FILE_PATH"] = f
+					runCmd(t, fmt.Sprintf("%s fs upload -s files/ro_1k lakefs://%s/%s/%s", Lakectl(), vars["REPO"], vars["BRANCH"], vars["FILE_PATH"]), false, false, vars)
+
+				}
+			}
+
+			// add local files
+			if len(tc.uncommittedLocal) > 0 {
```
Comment (N-o-Z) [inline esti/lakectl_local_test.go:584]: I'm not sure but do you need this check?

File: esti/lakectl_local_test.go [hunk 9]
```diff
@@ -464,6 +464,145 @@ func TestLakectlLocal_commit(t *testing.T) {
 	}
 }
 
+func TestLakectlLocal_commit_remote_uncommitted(t *testing.T) {
```
Comment (N-o-Z) [inline esti/lakectl_local_test.go:467]: Thanks for adding the tests! 💪🏽

File: cmd/lakectl/cmd/local_commit.go (hunk 1 above)
Comment (N-o-Z) [reply cmd/lakectl/cmd/local_commit.go:57]: I get it now 😅 
Can we please add the explanation in the code so it will be clear for anyone looking at it?

File: cmd/lakectl/cmd/local_commit.go [hunk 10]
```diff
@@ -56,6 +109,11 @@ var localCommitCmd = &cobra.Command{
 		resp, err := client.GetBranchWithResponse(cmd.Context(), remote.Repository, remote.Ref)
 		DieOnErrorOrUnexpectedStatusCode(resp, err, http.StatusOK)
 
+		hasChangesOutsideSyncedPrefix := hasExternalChange(cmd.Context(), client, remote, idx)
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/local_commit.go:112]: Optimally this should move to line 171 to reduce the chance of a race

**Review by N-o-Z — APPROVED** [review]
@eladlachmi Thanks for the great work.
Approved with a couple of comments.
It should be clear that this solution is not bulletproof and is susceptible to a race
